doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+  print_chunk out g seen fmt "Gallina The Vernacular" ["gallina"] ["tactic_expr5"];
+  print_chunk out g seen fmt "intropattern_list_opt" ["intropattern_list"; "or_and_intropattern_loc"] ["operconstr"; "tactic_expr5"];
+  print_chunk out g seen fmt "simple_tactic" ["simple_tactic"]
+    ["tactic_expr5"; "tactic_expr3"; "tactic_expr2"; "tactic_expr1"; "tactic_expr0"];
+
+  (*print_chunk out g seen fmt "Ltac" ["tactic_expr5"] [];*)
+  print_chunk out g seen fmt "Ltac" ["tactic_expr5"] ["tactic_expr4"];
+  print_chunk out g seen fmt "Ltac 4" ["tactic_expr4"] ["tactic_expr3"; "tactic_expr2"];
+  print_chunk out g seen fmt "Ltac 3" ["tactic_expr3"] ["tactic_expr2"];
+  print_chunk out g seen fmt "Ltac 2" ["tactic_expr2"] ["tactic_expr1"];
+  print_chunk out g seen fmt "Ltac 1" ["tactic_expr1"] ["tactic_expr0"];
+  print_chunk out g seen fmt "Ltac 0" ["tactic_expr0"] [];
+
+
+  print_chunk out g seen fmt "command" ["command"] [];
+  print_chunk out g seen fmt "vernac_toplevel" ["vernac_toplevel"] [];
+  print_chunk out g seen fmt "vernac_control" ["vernac_control"] []
+
+  (*
+    let ssr_tops = ["ssr_dthen"; "ssr_else"; "ssr_mpat"; "ssr_rtype"] in
+    seen := StringSet.union !seen (StringSet.of_list ssr_tops);
+
+    print_chunk out g seen fmt "ssrindex" ["ssrindex"] [];
+    print_chunk out g seen fmt "command" ["command"] [];
+    print_chunk out g seen fmt "binder_constr" ["binder_constr"] [];
+    (*print_chunk out g seen fmt "closed_binder" ["closed_binder"] [];*)
+    print_chunk out g seen fmt "gallina_ext" ["gallina_ext"] [];
+    (*print_chunk out g seen fmt "hloc" ["hloc"] [];*)
+    (*print_chunk out g seen fmt "hypident" ["hypident"] [];*)
+    print_chunk out g seen fmt "simple_tactic" ["simple_tactic"] [];
+    print_chunk out g seen fmt "tactic_expr" ["tactic_expr4"; "tactic_expr1"; "tactic_expr0"] [];
+    fprintf out "\n\nRemainder:\n";
+    print_in_order g (List.filter (fun x -> not (StringSet.mem x !seen)) g.order) StringSet.empty;
+  *)
+
+
+    (*seen := StringSet.diff !seen (StringSet.of_list ssr_tops);*)
+    (*print_chunk out g seen fmt "vernac_toplevel" ["vernac_toplevel"] [];*)
+
+let start_symbols = ["vernac_toplevel"; "vernac_control"; "tactic_mode"]
+let tokens = [ "BULLET"; "FIELD"; "IDENT"; "INT"; "STRING" ] (* don't report as undefined *)
+
+let report_bad_nts g file =
+  let rec get_nts refd defd bindings =
+    match bindings with
+    | [] -> refd, defd
+    | (nt, prods) :: tl ->
+      get_nts (List.fold_left (fun res prod ->
+                  StringSet.union res (StringSet.of_list (nts_in_prod prod)))
+                refd prods)
+        (StringSet.add nt defd) tl
+  in
+  let all_nts_ref, all_nts_def =
+    get_nts (StringSet.of_list tokens) (StringSet.of_list tokens) (NTMap.bindings g.map) in
+
+  let undef = StringSet.diff all_nts_ref all_nts_def in
+  List.iter (fun x -> warn "%s: Undefined symbol '%s'\n" file x) (StringSet.elements undef);
+
+  let reachable = StringSet.elements
```

Comment:
Don't create a list out of set when the only thing you need to check is membership! You're trading O(log n) for O(n) complexity.

Reply:
Perhaps an artifact of earlier refactorings.